### PR TITLE
[HunyuanImage3] Align system_prompt support with official implementation

### PR DESCRIPTION
**Pull request by skf-1999**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Add system prompt support for HunyuanImage to align with the official implementation. This PR introduces command line arguments --use-system-prompt and --system-prompt to allow users to specify different system prompt presets or custom prompts when generating images.
## Test Plan
New Arguments
| Argument              | Description               | Options                                                                                       | Default                                         |
| --------------------- | ------------------------- | --------------------------------------------------------------------------------------------- | ----------------------------------------------- |
| `--use-system-prompt` | System prompt preset type | `None`, `dynamic`, `en_vanilla`, `en_recaption`, `en_think_recaption`, `en_unified`, `custom` | `en_unified`                                    |
| `--system-prompt`     | Custom system prompt text | Any string                                                                                    | `None` (used when `--use-system-prompt=custom`) |

## Offline Inference:
```
python3 vllm-omni/examples/offline_inference/text_to_image/text_to_image.py \
  --model /data/HunyuanImage-3.0/ \
  --prompt "a cute cat" \
  --output dog.png \
  --num-inference-steps 50 \
  --guidance-scale 5.0 \
  --cfg-scale 4.0 \
  --seed 1234 \
  --tensor-parallel-size 8 \
  --use-system-prompt "en_unified"
```
## Online API (curl):
```
curl -X POST http://localhost:8091/v1/images/generations \
  -H "Content-Type: application/json" \
  -d '{
    "prompt": "A brown and white dog is running on the grass",
    "use_system_prompt": "en_unified",
    "num_inference_steps": 50,
    "n": 4,
    "size": "1024x1024",
    "seed": 1234
  }' | jq -r '.data[0].b64_json' | base64 -d > online_test.png
```
## Test Script Location: 
Requires two model weights: "tencent/HunyuanImage-3.0" and  "openai/clip-vit-base-patch32".
pytest tests/e2e/offline_inference/test_hunyuanimage3_text2img.py -v -s
CLIP Embedding Image Similarity Comparison
Test Configuration:
Prompt: "A brown and white dog is running on the grass"
Seed: 1234
Metric: CLIP Cosine Similarity between Baseline (Official Source) and PR Implementation

## Test Result:
<table>
  <thead>
    <tr>
      <th>System Prompt</th>
      <th>Baseline (Official)</th>
      <th>PR Implementation</th>
      <th>CLIP Score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>None<br><code>(seed=1234)</code></td>
      <td></td>
      <td></td>
      <td>0.994987</td>
    </tr>
    <tr>
      <td><code>en_recaption</code></td>
      <td></td>
      <td></td>
      <td>0.997418</td>
    </tr>
    <tr>
      <td><code>en_think_recaption</code></td>
      <td></td>
      <td></td>
      <td>0.999098</td>
    </tr>
    <tr>
      <td><code>en_vanilla</code></td>
      <td></td>
      <td></td>
      <td>0.998555</td>
    </tr>
    <tr>
      <td><code>en_unified</code></td>
      <td></td>
      <td></td>
      <td><b>0.999665</b></td>
    </tr>
    <tr>
      <td><code>dynamic</code><br><sup>(same as en_vanilla when bot_task=image)</sup></td>
      <td></td>
      <td></td>
      <td>0.998555</td>
    </tr>
  </tbody>
</table>

<blockquote>
<b>Note:</b> <code>dynamic</code> produces identical results to <code>en_vanilla</code> when <code>bot_task = image</code>.
</blockquote>

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/entrypoints/openai/api_server.py [hunk 1]
```diff
@@ -1286,7 +1286,13 @@ async def generate_images(request: ImageGenerationRequest, raw_request: Request)
         if request.negative_prompt is not None:
             prompt["negative_prompt"] = request.negative_prompt
         gen_params = OmniDiffusionSamplingParams(num_outputs_per_prompt=request.n)
-
+        extra_args = {}
+        if request.use_system_prompt is not None:
```
Comment (Bounty-hunter) [inline vllm_omni/entrypoints/openai/api_server.py:1288]: Better to put it into `prompt` ?

File: tests/test_system_prompt.py [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+    0.082096, 0.009610, -0.025717, 0.002566, -0.060849, 0.017403, 0.032650, 0.018658,
+    -0.030629, -0.025032, 0.005555, 0.000522, -0.009667, -0.043099, 0.005939, -0.027156,
+    0.045634, -0.011986, 0.002713, -0.032225, -0.015494, 0.028734, -0.036528, -0.033101,
+    -0.027174, 0.009490, -0.016537, 0.029435, 0.065709, -0.037711, -0.020497, -0.005578,
+    0.011768, 0.061035, -0.044676, 0.016113, -0.042945, -0.022579, 0.002430, 0.012474,
+    -0.018198, 0.030468, -0.016646, 0.019020, -0.035804, 0.001175, -0.018312, -0.010760,
+], dtype=torch.float32)
+
+LOCAL_CLIP_PATH = "openai/clip-vit-base-patch32"
+IMAGE_PATH= "/data/your_system_prompt_en_unified_picture"
```
Comment (Bounty-hunter) [inline tests/test_system_prompt.py]: why not add to ci, and it run automically.

File: vllm_omni/diffusion/models/hunyuan_image_3/system_prompt.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    # Use predefined English system prompts: vanilla (basic), recaption, think_recaption
+    elif sys_type in ['en_vanilla', 'en_recaption', 'en_think_recaption']:
+        return t2i_system_prompts[sys_type][0]
+    # Dynamic mode: automatically select system prompt based on bot_task type
+    elif sys_type == "dynamic":
+        # Think task: use chain-of-thought recaption prompt
+        if bot_task == "think":
+            return t2i_system_prompts["en_think_recaption"][0]
+        # Recaption task: use recaption prompt
+        elif bot_task == "recaption":
```
Comment (Fishermanykx) [inline vllm_omni/diffusion/models/hunyuan_image_3/system_prompt.py:199]: Do we need to support think & recaption in DiT stage? In my opinion think & recaption needs to be completed in AR stage.

File: vllm_omni/entrypoints/openai/api_server.py (hunk 1 above)
Comment (skf-1999) [reply vllm_omni/entrypoints/openai/api_server.py:1288]: Better not—system prompts bypass CFG dropping while user prompts participate in it, so behavior definitions should stay in system.

File: vllm_omni/diffusion/models/hunyuan_image_3/system_prompt.py (hunk 3 above)
Comment (skf-1999) [reply vllm_omni/diffusion/models/hunyuan_image_3/system_prompt.py:199]: Currently, only image is supported; other forms of bot-task are not supported.

File: tests/test_system_prompt.py (hunk 2 above)
Comment (skf-1999) [reply tests/test_system_prompt.py]: Very useful suggestion. I’ll adjust the code to make it automated.

File: tests/e2e/offline_inference/test_hunyuanimage3_text2img.py [hunk 4]
```diff
@@ -0,0 +1,3231 @@
...
+REPO_ROOT = Path(__file__).resolve().parents[1]
+STAGE_CONFIG_PATH = REPO_ROOT / "vllm_omni" / "model_executor" / "stage_configs" / "hunyuan_image_3_moe.yaml"
+
+pytestmark = [pytest.mark.advanced_model, pytest.mark.diffusion, pytest.mark.cuda, pytest.mark.gpu]
+
+# System prompt type. Options: None, dynamic, en_vanilla, en_recaption, en_think_recaption, en_unified
+# Below are the CLIP embedding tensors from the official HunyuanImage model (seed=1234, prompt: "A brown and white dog is running on the grass").
+# SEED_1234 denotes the output without system prompt, while the remaining entries correspond to outputs generated with different system prompts.
+
+SEED_1234 = torch.tensor(
```
Comment (gcanlin) [inline tests/e2e/offline_inference/test_hunyuanimage3_text2img.py:27]: @yenuo26 @congw729 Can we upload these tensors to remote repo and download them every time executing test to avoid introduce too many code lines?

File: vllm_omni/entrypoints/openai/api_server.py [hunk 5]
```diff
@@ -1264,7 +1264,13 @@ async def generate_images(request: ImageGenerationRequest, raw_request: Request)
...
+            extra_args["use_system_prompt"] = request.use_system_prompt
+        if request.system_prompt is not None:
+            extra_args["system_prompt"] = request.system_prompt
+        if extra_args:
+            gen_params.extra_args = extra_args
```
Comment (gcanlin) [inline vllm_omni/entrypoints/openai/api_server.py:1293]: This change is introducing a new field in generation API.  Before, we haven't collect all non-standard parameters into `extra_args`. If we plan to do that, should we also collect other non-standard parameters like `layers`, which is only for Qwen-Image-Layered? @SamitHuang @wtomin

File: vllm_omni/entrypoints/openai/protocol/images.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+# ruff: noqa: E501
```
Comment (gcanlin) [inline vllm_omni/entrypoints/openai/protocol/images.py]: Why do we add this line?

Comment (skf-1999) [reply vllm_omni/entrypoints/openai/protocol/images.py]: I'll fix that right away.

File: docs/design/feature/expert_parallel.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+cd examples/offline_inference/text_to_image
+python text_to_image.py \
+    --model Your-org/your-model \
+    --prompt "a cup of coffee on the table" \
+    --output "ep_enabled.png" \
+    --num-inference-steps 50 \
+    --guidance_scale 5.0 \
+    --tensor-parallel-size 8 \
+    --seed 1234 \
+    --enable_expert_parallel
```
Comment (gcanlin) [inline docs/design/feature/expert_parallel.md]: We should unify to `--enable-expert-parallel`

File: examples/online_serving/text_to_image/openai_chat_client.py [hunk 8]
```diff
@@ -45,6 +47,8 @@ def generate_image(
         lora_name: LoRA name (optional, defaults to path stem)
         lora_scale: LoRA scale factor (default: 1.0)
         lora_int_id: LoRA integer ID (optional, derived from path if not provided)
+        use_system_prompt: System prompt for generation. Use predefined types: 'en_unified', 'en_vanilla', 'en_recaption', 'en_think_recaption', 'dynamic', or 'None'; Or provide custom text string directly. Recommended en_unified.
```
Comment (gcanlin) [inline examples/online_serving/text_to_image/openai_chat_client.py]: It's better to add the usage of these fields in Hunyuan-Image doc.

File: vllm_omni/entrypoints/openai/protocol/images.py [hunk 9]
```diff
@@ -81,6 +81,14 @@ def validate_layers(cls, v):
 
     # vllm-omni extensions for diffusion control
     negative_prompt: str | None = Field(default=None, description="Text describing what to avoid in the image")
+    system_prompt: str | None = Field(
```
Comment (gcanlin) [inline vllm_omni/entrypoints/openai/protocol/images.py:84]: We may need a validator for this new field.

```
@field_validator('use_system_prompt')
def validate_system_prompt_type(cls, v):
    valid_types = ["None", "dynamic", "en_vanilla", "en_recaption", 
                   "en_think_recaption", "en_unified", "custom"]
    if v is not None and v not in valid_types:
        raise ValueError(f"Invalid use_system_prompt type: {v}")
    return v
```

Comment (skf-1999) [reply vllm_omni/entrypoints/openai/protocol/images.py:84]: Thank you for the suggestion, will revise promptly.

File: vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py [hunk 10]
```diff
@@ -992,10 +993,15 @@ def forward(
...
-        system_prompt: str | None = None,
         generator: torch.Generator | list[torch.Generator] | None = None,
         **kwargs,
     ) -> DiffusionOutput:
+        extra_args = getattr(req.sampling_params, "extra_args", {}) if hasattr(req, "sampling_params") else {}
+        use_system_prompt = extra_args.get("use_system_prompt")
+        system_prompt = extra_args.get("system_prompt")
+        if use_system_prompt is not None:
+            system_prompt = get_system_prompt(use_system_prompt, "image", system_prompt)
+            system_prompt = system_prompt.strip() if system_prompt is not None else ""
```
Comment (gcanlin) [inline vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py:1003]:
```suggestion
            system_prompt = system_prompt.strip() if system_prompt is not None else None
```

File: tests/e2e/offline_inference/test_system_prompt.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+from transformers import CLIPModel, CLIPProcessor
+
+from vllm_omni import Omni
+from vllm_omni.inputs.data import OmniDiffusionSamplingParams
+from vllm_omni.platforms import current_omni_platform
+
+PROMPT = "A brown and white dog is running on the grass"
+MODEL_NAME = "tencent/HunyuanImage-3.0"
+LOCAL_CLIP_PATH = "openai/clip-vit-base-patch32"
+REPO_ROOT = Path(__file__).resolve().parents[1]
```
Comment (gcanlin) [inline tests/e2e/offline_inference/test_system_prompt.py]:
```suggestion
REPO_ROOT = Path(__file__).resolve().parents[3]
```

**Comment by hsliuustc0106** [comment]
Good catch! This is a limitation of the current implementation.

**Current scope**: This PR focuses on image generation (bot_task="image") to align with the basic official functionality.

**Future work**: Support for "think" and "recaption" modes will be added in a follow-up PR.

Does the official HunyuanImage-3.0 implementation use these modes (think/recaption) dynamically? If so, we can prioritize adding this feature.

File: tests/e2e/offline_inference/test_hunyuanimage3_text2img.py (hunk 4 above)
Comment (yenuo26) [reply tests/e2e/offline_inference/test_hunyuanimage3_text2img.py:27]: Maybe we can later plan a dedicated directory to store these files.

File: tests/e2e/offline_inference/test_system_prompt.py [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+STAGE_CONFIG_PATH = REPO_ROOT / "vllm_omni" / "model_executor" / "stage_configs" / "hunyuan_image_3_moe.yaml"
+
+pytestmark = [pytest.mark.advanced_model, pytest.mark.diffusion, pytest.mark.cuda, pytest.mark.gpu]
```
Comment (gcanlin) [inline tests/e2e/offline_inference/test_system_prompt.py]: @yenuo26 Could you help check whether this mark is appropriate? It seems that `pytest.mark.cuda, pytest.mark.gpu` are redundant. Should it be like:

```suggestion
pytestmark = [pytest.mark.advanced_model, pytest.mark.diffusion]
```

File: docs/design/feature/expert_parallel.md [hunk 13]
```diff
@@ -0,0 +1,221 @@
+# Expert Parallel
```
Comment (lishunyang12) [inline docs/design/feature/expert_parallel.md]: This file is unrelated to the system prompt feature. Please split it into a separate PR — mixing unrelated changes makes review and bisect harder.

File: examples/offline_inference/text_to_image/text_to_image.py [hunk 14]
```diff
@@ -242,6 +242,20 @@ def parse_args() -> argparse.Namespace:
         action="store_true",
         help="Enable logging of diffusion pipeline stats.",
     )
+    parser.add_argument(
+        "--use-system-prompt",
+        type=str,
+        default=None,
+        help=(
+            "System prompt for generation. Use predefined types: 'en_unified', 'en_vanilla', 'en_recaption', 'en_think_recaption', 'dynamic', or 'None'; Or provide custom text string directly. Recommended en_unified. "
```
Comment (lishunyang12) [inline examples/offline_inference/text_to_image/text_to_image.py]: Nit: `--use-system-prompt` accepts any string but only a handful of values are valid. Consider using `choices=[...]` so argparse catches typos.

```suggestion
    parser.add_argument(
        "--use-system-prompt",
        type=str,
        default=None,
        choices=["None", "dynamic", "en_vanilla", "en_recaption",
                 "en_think_recaption", "en_unified", "custom"],
        help="System prompt preset for generation. Recommended: en_unified.",
    )
```

File: vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py [hunk 15]
```diff
@@ -992,10 +993,15 @@ def forward(
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py:990]: This line is a bit convoluted — `hasattr` guard + `getattr` default do the same thing. Simpler:

```suggestion
        extra_args = getattr(getattr(req, "sampling_params", None), "extra_args", {}) or {}
```

File: vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py (hunk 10 above)
Comment (skf-1999) [reply vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py:1003]: This part we should keep same with official hunyuan: in HunyuanImage-3.0/hunyuan_image_3/modeling_hunyuan_image_3.py system_prompt = system_prompt.strip() if system_prompt is not None else ""

**Comment by skf-1999** [comment]
> Good catch! This is a limitation of the current implementation.
> 
> **Current scope**: This PR focuses on image generation (bot_task="image") to align with the basic official functionality.
> 
> **Future work**: Support for "think" and "recaption" modes will be added in a follow-up PR.
> 
> Does the official HunyuanImage-3.0 implementation use these modes (think/recaption) dynamically? If so, we can prioritize adding this feature.

According to the official documentation, these modes are supported, but I haven't fully verified this with the official implementation yet (may require HunyuanImage-3.0-Instruct weights).

File: vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py (hunk 15 above)
Comment (skf-1999) [reply vllm_omni/diffusion/models/hunyuan_image_3/pipeline_hunyuan_image_3.py:990]: Thanks for the suggestion, addressed.

**Comment by gcanlin** [comment]
@hsliuustc0106 I dismissed your `requested change` review because all issues you mentioned have been fixed. Let's wait CI happy and merge this PR ASAP.

**Comment by congw729** [comment]
@gcanlin @hsliuustc0106  All CI passed. This PR is okay for merge.
